crates/torii/sqlite/src/cache.rs

```diff
@@ -147,3 +154,98 @@ impl LocalCache {
         self.token_id_registry.write().await.insert(token_id);
     }
 }
+
+#[derive(Debug)]
+pub struct ContractClassCache<P: Provider + Sync + std::fmt::Debug> {
+    pub classes: RwLock<HashMap<Felt, (Felt, ContractClass)>>,
+    pub provider: Arc<P>,
+}
+
+impl<P: Provider + Sync + std::fmt::Debug> ContractClassCache<P> {
+    pub fn new(provider: Arc<P>) -> Self {
+        Self { classes: RwLock::new(HashMap::new()), provider }
+    }
+
+    pub async fn get(
+        &self,
+        contract_address: Felt,
+        mut block_id: BlockId,
+    ) -> Result<ContractClass, Error> {
+        {
+            let classes = self.classes.read().await;
+            if let Some(class) = classes.get(&contract_address) {
+                return Ok(class.1.clone());
+            }
+        }
+
+        let class_hash = match self.provider.get_class_hash_at(block_id, contract_address).await {
+            Ok(class_hash) => class_hash,
+            Err(e) => match e {
+                // if we got a block not found error, we probably are in a pending block.
+                ProviderError::StarknetError(StarknetError::BlockNotFound) => {
+                    block_id = BlockId::Tag(BlockTag::Pending);
+                    self.provider.get_class_hash_at(block_id, contract_address).await?
+                }
+                _ => return Err(Error::ProviderError(e)),
+            },
+        };
+        let class = self.provider.get_class_at(block_id, contract_address).await?;
+        self.classes.write().await.insert(contract_address, (class_hash, class.clone()));
```

Comment:
If it mostly used for ABI lookup, storing the ABI only instead of the whole class with byte code could result in less memory being used.

Also, if the usage is mostly mapping a selector to a name, the mapping could be done the first time the class is fetched instead of storing the raw ABI.

Reply:
Oh nice catch yes!